In-place receive for unstructured

bettiolm opened the pull request.

mbianco commented on include/ghex/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,564 @@
+#ifndef INCLUDED_GHEX_COMMUNICATION_OBJECT_IPR_HPP
+#define INCLUDED_GHEX_COMMUNICATION_OBJECT_IPR_HPP
+
+// TO DO: remove unneeded includes
```
Check this before merging

bettiolm replied:
Done, added a couple of issues as well.

mbianco commented on include/ghex/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,564 @@
+
+    namespace ghex {
+
+        // namespace unstructured {
```
remove? Ho do we protect against mis-use?

bettiolm replied:
Done, used a specialization for unstructured.

mbianco commented on include/ghex/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,564 @@
+                    /** @brief pair of domain ids + tag id with ordering */
+                    struct domain_id_pair_and_tag {
+
+                            domain_id_type first_id;
```
identation

bettiolm replied:
Done

mbianco commented on tests/unstructured_user_concepts.cpp:
```diff
@@ -486,13 +522,87 @@ TEST(unstructured_user_concepts, data_descriptor) {
     EXPECT_NO_THROW(co.bexchange(patterns(data)));
 
     auto h = co.exchange(patterns(data));
+    //auto h = co.exchange_ipr(patterns(data));
```
```suggestion
```

bettiolm replied:
Done

mbianco reviewed (commented):
I'm very slow on this one. Some  minor comments for now

mbianco commented on include/ghex/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,564 @@
+                    };
+
+                    /** @brief Holds a pointer to a set of iteration spaces and a callback function pointer
+                     * which is used to store a field's pack or unpack member function.
```
are these ones used for unpack? I guess not.

bettiolm replied:
This struct is actually used for the receive part as well, using `std::function<void()>` as function type. I first thought of not using a func at all, but this requires at least to define a new struct (or a specialization of this one), and I think that keeping it as it is is less confusing in the end. I'll mention this in one of the two opened issues anyway.

bettiolm replied:
I changed the code a little bit here now. `field_info` is actually used only for the send part, so only with pack function. I added a specific struct for the receive, without callback (unpack) function.

mbianco commented on include/ghex/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,564 @@
+                     * The type-erased field_ptr member is only used for the gpu-vector-interface. // TO DO: verify is still needed in this case
+                     * @tparam Function either pack or unpack function pointer type */
+                    template<typename Function>
+                    struct field_info {
```
Also this struct is the same as the co_2. Would it make sense to use some composition to reuse some code? (how much code would be saved by doing so?)

bettiolm replied:
Moved to refactoring issue.

mbianco commented on include/ghex/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,564 @@
+                    /** @brief Holds serial buffer memory and meta information associated with it.
+                     * @tparam Vector contiguous buffer memory type
+                     * @tparam Function pack function pointer type (not used for unpacking)*/
+                    template<class Vector, class Function>
```
This struct also is the same as co_2. Maybe instead of composing classes, we can put these sub-structs into an impl namespace in a separate file and use them when we need them

bettiolm replied:
Moved to refactoring issue.

boeschf commented on include/ghex/unstructured/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,573 @@
+                    using grid_type = unstructured::detail::grid<index_type>;
+                    using domain_id_type = DomainIdType;
+
+                    // friend class // TO DO: is friend needed?
```
the friend is needed because of private ctor

bettiolm replied:
Right. Done.

boeschf commented on include/ghex/unstructured/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,573 @@
+                    friend class communication_object_ipr<communicator_type,grid_type,domain_id_type>;
+
+                    // members
+                    communicator_type m_comm; // TO DO: is it needed?
```
comm is probably not needed

bettiolm replied:
Removed from class members.

boeschf commented on include/ghex/unstructured/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,573 @@
+                        });
+                    }
+
+                    // TO DO: double check from here on. First of all, some code could be avoided
```
is this resolved?

bettiolm replied:
Yes, I just left it behind

bettiolm commented on include/ghex/unstructured/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,561 @@
+                            Vector buffer;
+                            std::size_t size;
+                            std::vector<field_info_type> field_infos;
+                            cuda::stream m_cuda_stream;
```
Is it ok to stick with this naming convention (same used in `communication_object_2`)?

mbianco replied:
I did not notice the names here. We use `m_` for members that are not assumed to be accessed publicly.
Anyway, since these members are not accessed in this PR, I would leave the code as is, since otherwise you have to make more diffs.

bettiolm replied:
Ok, sure

bettiolm commented on include/ghex/unstructured/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,561 @@
+                            ipr_message message;
+                            std::size_t size;
+                            std::vector<field_info_type> field_infos;
+                            cuda::stream m_cuda_stream;
```
Same comment here on the naming convention

mbianco replied:
as above

bettiolm replied:
Yep

bettiolm commented on include/ghex/unstructured/communication_object_ipr.hpp:
```diff
@@ -0,0 +1,561 @@
+                                        p1.second.message.resize(p1.second.size);
+                                        m.m_recv_futures.emplace_back(
+                                            typename std::remove_reference_t<decltype(m)>::future_type{
+                                                m_comm.recv(p1.second.message, p1.second.address, p1.second.tag).m_handle
```
Here I used the same constructor as in `communication_object_2`, i.e. the one with `handle` as argument, which moves only the handle. Not a big deal, but I think that using directly the future move constructor could also be an option, since we are using the same future specialization in both cases anyway. Probably this could be a bit too strict if one day we should decide (for some reason) not to use the `future<void>` specialization as return type for the send / recv functions.

mbianco replied:
I'm not sure I understand completely. The trade-off between reuse of patterns and code is tricky. I would not spend too much thinking about possible future choices, and maybe put a comment explaining what to do if an error appears about it

bettiolm replied:
Ok, I just added a comment in the code

bettiolm commented:
I addressed the comments and added a couple of issues (mostly minor refactoring) which can be addresses after the merge in my opinion. From my side it is ok now.

mbianco commented on include/ghex/unstructured/communication_object_ipr.hpp:
```diff
@@ -376,7 +386,7 @@ namespace gridtools {
                              typename Field,
                              typename O>
-                    void allocate(Memory& mem,
+                    void allocate(Memory& mem, // reference not strictly needed here
```
why is this comment useful? could you expand it?

bettiolm replied:
It's just a reminder, memory is already a pointer, and it is used as such inside, so a copy could also be ok. Same happens in comm2. I just did not want to change much the interface with respect of the comm2 one, since that should be harmless.

bettiolm replied:
I pushed some documentation and removed the pass by reference from this function (allocate), in which it is actually not needed.

bettiolm replied:
From my side it should be ok now

mbianco commented on include/ghex/unstructured/communication_object_ipr.hpp:
```diff
@@ -410,23 +417,17 @@ namespace gridtools {
-                             typename Pool,
                              typename Field = void>
-                    void set_recv_size(Memory& memory,
+                    void set_recv_size(Memory& memory, // reference not strictly needed here
```
why is this comment useful? could you expand it?

bettiolm replied:
as before

bettiolm replied:
Actually not, sorry. Here memory is not a pointer as before, and represents the recv memory. Reference is needed. I pushed the updated version.

mbianco commented on include/ghex/unstructured/communication_object_ipr.hpp:
```diff
@@ -473,14 +474,13 @@ namespace gridtools {
                              typename Pool,
                              typename Field = void>
-                    void allocate_send(Memory& memory,
+                    void allocate_send(Memory& memory, // reference not strictly needed here
```
why is this comment useful? could you expand it?

bettiolm replied:
as before

bettiolm replied:
Same as for `set_recv_size`
